arduino/smartcar/smartcar.ino

```diff
@@ -14,18 +18,77 @@ void setup() {
 }
 
 void loop() {
-  car.setSpeed(90);
-  preventCrash();
+  car.setSpeed(50); 
+  preventCrashWithUltrasonic(); 
+  preventCrashWithInfraredFront(); 
+  preventCrashWithInfraredLeft();
+  preventCrahsWithInfraredRight();
+  preventCrashWithInfraredBack();
 }
 
-void preventCrash() {
-  unsigned int fDistance = front.getDistance();
+void preventCrashWithUltrasonic() {
+  unsigned int fDistance = frontUS.getDistance();
   if (fDistance > 0 && fDistance < 200) {
+    Serial.print("Obstacle detected by front ultrasonic sensor");
     car.setSpeed(0);
     delay(500);
-    control.overrideMotorSpeed(100, -100);
+    control.overrideMotorSpeed(50, -100);
     delay(800);
     car.setSpeed(0);
     delay(200);
   }
 }
+
+void preventCrashWithInfraredFront() {
+  unsigned int fDistance = frontIR.getDistance();
+  if (fDistance > 0 && fDistance < 200) {
+    Serial.print("Obstacle detected by front infrared sensor"); 
+    car.setSpeed(0);
+    delay(500);
+    control.overrideMotorSpeed(50, -100);
+    delay(800);
+    car.setSpeed(0);
+    delay(200);
+  }
+}
+
+void preventCrashWithInfraredLeft() {
+  unsigned int fDistance = leftIR.getDistance();
+  if (fDistance > 0 && fDistance < 200) {
+    Serial.print("Obstacle detected by left infrared sensor"); 
+    car.setSpeed(0);
+    delay(500);
+    control.overrideMotorSpeed(50, -100);
+    delay(800);
+    car.setSpeed(0);
+    delay(200);
+  }
+}
+
+void preventCrahsWithInfraredRight() {
+  unsigned int fDistance = rightIR.getDistance();
+  if (fDistance > 0 && fDistance < 200) {
+    Serial.print("Obstacle detected by right infrared sensor"); 
+    car.setSpeed(0);
+    delay(500);
+    control.overrideMotorSpeed(50, -100);
+    delay(800);
+    car.setSpeed(0);
+    delay(200);
+  }
+}
+
+void preventCrashWithInfraredBack() {
+  unsigned int fDistance = backIR.getDistance();
+  if (fDistance > 0 && fDistance < 200) {
+    Serial.print("Obstacle detected by back infrared sensor"); 
+    car.setSpeed(0);
+    delay(500);
+    control.overrideMotorSpeed(50, -100);
+    delay(800);
+    car.setSpeed(0);
+    delay(200);
+  }
+}
```

Comment:
There is a lot of duplicated code but object oriented programming can save the day. :)
All sensors that measure distance, derive from the [DistanceSensor](https://platisd.github.io/smartcar_shield/interface_distance_sensor.html) interface. So what if you use that instead as an argument?

For example, all those methods could be replaced with:

```cpp
void preventCrashWithInfrared(DistanceSensor& distanceSensor) {
  unsigned int fDistance = distanceSensor.getDistance();
  // The rest remains the same...
}
```

Reply:
Thanks, the patch is updated. I removed the duplicated code.